src/v/cluster/rm_stm.cc

```diff
@@ -2726,4 +2731,21 @@ ss::future<> rm_stm::clear_old_tx_pids() {
       });
 }
 
+ss::future<> rm_stm::clear_old_idempotent_pids() {
+    while (_log_state.lru_idempotent_pids.size() > 100500) {
+        auto pid_for_delete = _log_state.lru_idempotent_pids.front().entry.pid;
+        auto rw_lock = get_idempotent_producer_lock(pid_for_delete);
+        auto lock = rw_lock->try_write_lock();
+        if (lock) {
+            _log_state.lru_idempotent_pids.pop_front();
+            _log_state.seq_table.erase(pid_for_delete);
+            _inflight_requests.erase(pid_for_delete);
+            rw_lock->write_unlock();
+            _idempotent_producer_locks.erase(pid_for_delete);
```

Comment:
this is tricky, imagine that there is a read lock waiting for the write lock to be over, maybe as long as we invoke `write_unlock` the lock grants read access and if it's true then there is a risk of removing the rwlock while it's being used.